packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts

```diff
@@ -67,30 +71,29 @@ export const useGridSelection = (apiRef: GridApiRef): void => {
 
       logger.debug(`Selecting row ${id}`);
 
-      const allowMultiSelect =
-        allowMultipleOverride || allowMultipleSelectionKeyPressed.current || checkboxSelection;
+      setGridState((state) => {
+        let selectionState: GridSelectionState = { ...state.selection };
+        const allowMultiSelect =
+          allowMultipleOverride ||
+          (!disableMultipleSelection && isMultipleKey) ||
+          checkboxSelection;
 
-      if (allowMultiSelect) {
-        setGridState((state) => {
-          // eslint-disable-next-line prefer-object-spread
-          const selectionState: GridSelectionState = Object.assign({}, state.selection);
+        if (allowMultiSelect) {
           const isRowSelected =
             !allowMultiSelect || isSelected == null ? selectionState[id] === undefined : isSelected;
-
           if (isRowSelected) {
             selectionState[id] = id;
           } else {
             delete selectionState[id];
           }
-          return { ...state, selection: selectionState };
-        });
-      } else {
-        setGridState((state) => {
-          const selectionState: GridSelectionState = {};
+        } else if (disableMultipleSelection && isMultipleKey && selectionState[id] !== undefined) {
+          selectionState = {};
+        } else {
+          selectionState = {};
           selectionState[id] = id;
```

Comment:
You were not honoring `isSelected`. I think we can simplify this block with:

```suggestion
        } else {
          const isRowSelected =
            isSelected == null ? !isMultipleKey || selectionState[id] === undefined : isSelected;
          selectionState = {};
          if (isRowSelected) {
            selectionState[id] = id;
          }
```

Reply:
Updated